nginx/ngx_http_js_module.c

```diff
@@ -7707,43 +7710,48 @@ ngx_http_js_init(ngx_conf_t *cf)
     ngx_http_top_body_filter = ngx_http_js_body_filter;
 
     jmcf = ngx_http_conf_get_module_main_conf(cf, ngx_http_js_module);
-    jlcf = ngx_http_conf_get_module_loc_conf(cf, ngx_http_js_module);
 
-    if (jlcf->engine == NULL) {
-        found_issue = 0;
+    found_issue = 0;
 
-        if (jmcf->periodics != NULL) {
-            periodic = jmcf->periodics->elts;
+    if (jmcf->periodics != NULL) {
+        periodic = jmcf->periodics->elts;
 
-            for (i = 0; i < jmcf->periodics->nelts; i++) {
+        for (i = 0; i < jmcf->periodics->nelts; i++) {
+            jlcf = ngx_http_get_module_loc_conf(periodic[i].conf_ctx,
+                                                ngx_http_js_module);
+            if (jlcf->engine == NULL) {
                 ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
                               "no \"js_import\" directives found for "
                               "\"js_periodic\" \"%V\" in %s:%ui",
                               &periodic[i].method, periodic[i].file_name,
                               periodic[i].line);
+                found_issue = 1;
             }
-
-            found_issue = 1;
         }
+    }
 
-        cmcf = ngx_http_conf_get_module_main_conf(cf, ngx_http_core_module);
-        key = cmcf->variables_keys->keys.elts;
+    cmcf = ngx_http_conf_get_module_main_conf(cf, ngx_http_core_module);
+    key = cmcf->variables_keys->keys.elts;
 
-        for (i = 0; i < cmcf->variables_keys->keys.nelts; i++) {
-            v = key[i].value;
-            if (v->get_handler == ngx_http_js_variable_set) {
-                data = (ngx_js_set_t *) v->data;
+    for (i = 0; i < cmcf->variables_keys->keys.nelts; i++) {
+        v = key[i].value;
+        if (v->get_handler == ngx_http_js_variable_set) {
```

Comment:
TBH, I don't think its worth having all this validation code if runtime check is still required. So I'd suggest to remove config time validation, like it was in previous revision (but for both js_set and periodics). We really need to check variable usage place not the declaration, but Nginx doesn't provide such hooks...

Besides, the js_set change can break some existing code.